README


This archive contains the distribution Locale-Simple,
version 0.018:

  Functions for translate text based on gettext data, also in JavaScript

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>, Torsten Raudssus <torsten@raudss.us>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
==================================================
Changes from 1740-11-23 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.018 at 2014-09-07 17:22:38 +0000
------------------------------------------

  Change: 88e9c5a2aa5f7897ecc98950484859c2e8a69c89
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2014-09-07 19:22:18 +0000

    Added old (different working) scraper version 

  Change: 51c0f2442f2d22b7190fa9e735854abd7a33f8f1
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-09-06 23:18:40 +0000

    allow for multiple concat ops and add xslate concat op ~ 

  Change: fac3922df2f924741b97730489cf02f1d4e4b694
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-21 01:33:34 +0000

    new travis settings 

  Change: 3a2e6bd93acabf9a841812d49441e27485fb1920
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-21 02:01:22 +0000

    all the test data needs to be treated by dzil as binary files 

  Change: d354738a61681b1adaa228c9b99105c1be77eb35
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:59:07 +0000

    empty merge commit to aid history spelunking 

  Change: 5bbbce110bfdc22487f8a277cce7691f37c0a3d0
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    add option to suppress line number output 

  Change: c77da4bcc2ee7444d27b08ec2785d5b117670575
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    ignore dzil build artefacts 

  Change: 54abf6cc9ea0b2a890246c4b5ff4d3552f657dc0
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    concatenated dynamic strings can contain function calls 

  Change: fda3c3e96796b464298f65be13483383aac0bdab
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    perl 5.10 needs parens on that shift 

  Change: d2b351477b816a68daf6d4533c749acadc873e0a
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    force the scraper to always process files in the same order 

  Change: ad424870b8e55259f9ef620151171d7bcfabe7c5
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:51 +0000

    ensure tests work in release configuration and after dzil munging 

  Change: ab274636dff23eefc2f3812141b94a4622e0cb6e
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    move expect_op to Parser.pm, as it is specific to LSSP 

  Change: 2f4fc01d9ce9f1fef077a115ec6565b1faac21dd
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    add abstracts for the newly added modules 

  Change: 73679cc14493df6827650d389879d45285ddb8e9
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    dzil config now uses AutoPrereqs 

  Change: c88f9150a8ac04b54d732e7d84a65bc3eb05402b
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    expect_op now handles preceding whitespace transparently 

  Change: 1cc552da4524ab6fba73509c8631ee003524382b
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    reports for where tokens are found now use the line number of the l*
    call 

  Change: 106f4c0a6d0ee116077227f2f73d9fba813171a5
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    make arguments() semantically more clean by moving result saving to
    call() 

  Change: 38e62bea8996620911ccf101f4fdc1a15b193550
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    don't strip error line numbers in tests, report doesn't print them
    anymore 

  Change: 241d96d976237742def3511af29d8e57aa897027
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:50 +0000

    improve readability of arguments sub 

  Change: e9c05a529f0f86eb368aa36e3da8d4b788f2d2b7
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    converted required_args methods into one method 

  Change: aae77a545878f8e2e0f6c6d5fcb6213abdfad55a
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    expect_string is now expect_op, which never returns the op 

  Change: 04161839133a2f1c26d999727907a71c8a054e11
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    introduce with_ws to execute a parse while keeping whitespace 

  Change: 0743bd2c0bfc5e0cceb3abd2d0d7e3f47b238373
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    introduce shortcuts for some curries, fixes list_of curry
    incompatibility 

  Change: 198fa19ac24ff9b27b5e54b704348e07fe947768
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    improve readability by liberal application of curry 

  Change: ede065b3109d7903f6ef76584909ce2aacb7abed
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    improve readability of parser code by factoring out generic shortcuts 

  Change: 9f46c2be62b444c7ccc9a1f871a5e5958ed47180
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    improve readability of parser code by renaming and refactoring 

  Change: cc052d2d718cd54e5e3d1ae16952d2257678f542
  Author: Christian Walde <walde.christian@googlemail.com>
  Date : 2014-02-20 18:57:49 +0000

    converted extra-arg parsing from maybe to list, fixing extra arg
    counts 

  Change: ba349b9ffd92664044f1736d9c7485b4666d1505
  Author: Christian Walde <walde.christian@googlemail.com>

lib/Locale/Simple.pm
package Locale::Simple;
BEGIN {
  $Locale::Simple::AUTHORITY = 'cpan:GETTY';
}
# ABSTRACT: Functions for translate text based on gettext data, also in JavaScript
$Locale::Simple::VERSION = '0.018';
use strict;
use warnings;

use Exporter 'import';
use Locale::TextDomain 1.23 (); # to ensure we get the right version of gettext_dumb
use Locale::gettext_dumb qw(:locale_h :libintl_h);
use POSIX qw' setlocale ';
use IO::All 0.41 -utf8;

our @EXPORT = qw(
	
	l_dir
	l_lang
	l_dry
	l_nolocales

	l
	ln
	lp
	lnp
	ld
	ldn
	ldp
	ldnp
	
	ltd

);

my $dry;
my $nowrite;
my $nolocales;

my %tds;
my $dir;

sub l_nolocales { $nolocales = shift }
sub l_dry { $dry = shift; $nowrite = shift; $nolocales = 1 if $dry }

sub gettext_escape {
	my ( $content ) = @_;
	$content =~ s/\\/\\\\/g;
	$content =~ s/\n/\\n/g;
	$content =~ s/"/\\"/g;
	return $content;
}

sub sprintf_compare {
	my ( $first, $second ) = @_;
	my $re = qr/(?:%%|%(?:[0-9]+\$)?[+-]?(?:[ 0]|\'.)?-?[0-9]*(?:\.[0-9]+)?[bcdeufFosxX])/;
	my @placeholder_first = sort { $a cmp $b } map { my @chars = split(//,$_); pop @chars; } ($first =~ m/$re/g);
	my @placeholder_second = sort { $a cmp $b } map { my @chars = split(//,$_); pop @chars; } ($second =~ m/$re/g);
	return join("",@placeholder_first) eq join("",@placeholder_second);
}

sub coderef_hash {{
	l => sub { l(@_) },
	ln => sub { ln(@_) },
	lp => sub { lp(@_) },
	lnp => sub { lnp(@_) },
	ld => sub { ld(@_) },
	ldn => sub { ldn(@_) },
	ldp => sub { ldp(@_) },
	ldnp => sub { ldnp(@_) },
}}

sub l_dir { $dir = shift }

sub l_lang {
	my $primary = shift;
	$ENV{LANGUAGE} = $primary;
	$ENV{LANG} = $primary;		# unsure if these ENV assignments are still needed with the setlocale below
	$ENV{LC_ALL} = $primary;
	$ENV{LC_MESSAGES} = $primary; # set locale for messages if the system supports it
	setlocale( LC_MESSAGES, $primary ) if eval { LC_MESSAGES }; # set locale for messages if the system supports it
}

# write dry
sub wd { io($dry)->append(join("\n",@_)."\n\n") if !$nowrite }

# l(msgid,...)
sub l { return ldnp('',undef,shift,undef,undef,@_) }
# ln(msgid,msgid_plural,count,...)
sub ln { return ldnp('',undef,@_) }
# lp(msgctxt,msgid,...)
sub lp { return ldnp('',shift,shift,undef,undef,@_) }
# lnp(msgctxt,msgid,msgid_plural,count,...)
sub lnp { return ldnp('',shift,shift,shift,shift,@_) }
# ld(domain,msgid,...)
sub ld { return ldnp(shift,undef,shift,undef,undef,@_) }
# ldn(domain,msgid,msgid_plural,count,...)
sub ldn { return ldnp(shift,undef,shift,shift,shift,@_) }
# ldp(domain,msgctxt,msgid,...)
sub ldp { return ldnp(shift,shift,shift,undef,undef,@_) }
# ldnp(domain,msgctxt,msgid,msgid_plural,count,...)
sub ldnp {
	die "please set a locale directory with l_dir() before using other translate functions" unless $dir || $nolocales;
	my ($td, $ctxt, $id, $idp, $n) = (shift,shift,shift,shift,shift);
	my @args = @_;
	unshift @args, $n if $idp;
	my $return;
	if ($dry) {
		if (!$nowrite) {
			my @save;
			push @save, '# domain: '.$td if $td;
			push @save, 'msgctxt "'.gettext_escape($ctxt).'"' if $ctxt;
			push @save, 'msgid "'.gettext_escape($id).'"';
			push @save, 'msgid_plural "'.gettext_escape($idp).'"' if $idp;
			wd(@save);
		}
		$return = sprintf($idp && $n != 1 ? $idp : $id, @args);
	} else {
		# L::TD handles msg ids as bytes internally
		utf8::encode($id);
		my $gt = dnpgettext($td, $ctxt, $id, $idp, $n);
		# Fixing bad utf8 handling
		utf8::decode($gt);
		$return = sprintf($gt,@args);
	}
	return $return;
}

sub ltd {
	die "please set a locale directory with l_dir() before using other translate functions" unless $dir || $nolocales;
	my $td = shift;
	unless (defined $tds{$td}) {
		bindtextdomain($td,$dir);
		bind_textdomain_codeset($td,'utf-8');
		$tds{$td} = 1;
	}
	textdomain($td);
}

1;

__END__

=pod

=head1 NAME

Locale::Simple - Functions for translate text based on gettext data, also in JavaScript

=head1 VERSION

version 0.018

=head1 SYNOPSIS

  use Locale::Simple;

  l_dir('data/locale');
  ltd('test');
  l_lang('de_DE');

  print l("Hello"); # "Hallo"
  print ln("You have %d message","You have %d messages",4); # 'Du hast 4 Nachrichten'

Usage in JavaScript:

  <script language="javascript" src="../../share/js/sprintf.js"></script>
  <script language="javascript" src="../../share/js/gettext/Gettext.js"></script>
  <script language="javascript" src="../../share/js/locale_simple.js"></script>
  <script language="javascript" src="locale/de_DE/LC_MESSAGES/test.json"></script>

  ltd('test');

  l("Hello");
  ln("You have %d message","You have %d messages",4);

Sample PO file, in this case data/locale/test.po

  msgid ""
  msgstr ""
  "Language: de_DE\n"
  "MIME-Version: 1.0\n"
  "Content-Type: text/plain; charset=UTF-8\n"
  "Content-Transfer-Encoding: 8bit\n"
  "Plural-Forms: nplurals=2; plural=n != 1;"

  msgid "You have %d message"
  msgid_plural "You have %d messages"
  msgstr[0] "Du hast %d Nachricht"
  msgstr[1] "Du hast %d Nachrichten"

  msgid "Hello"
  msgstr "Hallo"

=head1 DESCRIPTION

This is a small wrapper around Gettext functionality that integrates sprintf and makes
it a bit more easy to setup the internationalization. It ONLY supports UTF8 data, and in
or output, that is a fixed setup (and always will be).

Gettext in Perl requires compiled po files, so called mo files. You can generate those
with the following command (if you have gettext in general installed on your system):

  msgfmt -o data/locale/test.mo data/locale/test.po

The Gettext implementation in JavaScript which is wrapped, requires a json file to be
generated out of the po. This can be achieved with po2json which is delivered with
this package. Sadly it only generates the json and doesnt integrate it into the
translation storage in the JavaScript. To generate this you can do:

  echo -n "locale_data['test'] = " >data/locale/test.json
  po2json data/locale/test.po >>data/locale/test.json
  echo ";" >>data/locale/test.json

B<WARNING> it could be that the way how to integrate this in JavaScript might change
in future version. Please check this place here on every upgrade for further informations.

=encoding utf8

=head1 SEE ALSO

=head2 L<Locale::Messages>

=head1 SUPPORT

Repository

  http://github.com/Getty/p5-locale-simple
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-locale-simple/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>, Torsten Raudssus <torsten@raudss.us>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Locale/Simple/Scraper.pm
use strict;
use warnings;

package Locale::Simple::Scraper;
BEGIN {
  $Locale::Simple::Scraper::AUTHORITY = 'cpan:GETTY';
}
$Locale::Simple::Scraper::VERSION = '0.018';
# ABSTRACT: scraper to find translation tokens in a directory

use Exporter 'import';
use Getopt::Long;
use File::Find;
use Cwd;
use Locale::Simple;
use Data::Dumper;
use Locale::Simple::Scraper::Parser;

our @EXPORT = qw(scrape);

sub scrape {
    @ARGV = @_;

    $| = 1;

    # Supported filetypes:
    my $js_ext = "";    # Javascript
    my $pl_ext = "";    # Perl
    my $py_ext = "";    # Python
    my $tx_ext = "";    # Text::Xslate (Kolon or Metakolon)

    my @ignores;
    my @only;

    my $output = 'po';
    my ($md5, $no_line_numbers);

    GetOptions(
        "js=s"            => \$js_ext,
        "pl=s"            => \$pl_ext,
        "py=s"            => \$py_ext,
        "tx=s"            => \$tx_ext,
        "ignore=s"        => \@ignores,
        "only=s"          => \@only,
        "output=s"        => \$output,
        "md5"             => \$md5,
        "no_line_numbers" => \$no_line_numbers,
    );

    # could add Getopt::Long here for override

    my @js = split( ",", $js_ext );
    push @js, 'js';

    my @pl = split( ",", $pl_ext );
    push @pl, 'pl', 'pm', 't';

    my @tx = split( ",", $tx_ext );
    push @tx, 'tx';

    my @py = split( ",", $py_ext );
    push @py, 'py';

    # extension list
    my %e = (
        ( map { $_ => 'js' } @js ),
        ( map { $_ => 'pl' } @pl ),
        ( map { $_ => 'tx' } @tx ),
        ( map { $_ => 'py' } @py ),
    );

    # functions with count of locale simple with function of parameter
    #
    # 1 = msgid
    # 2 = msgid_plural
    # 3 = msgctxt
    # 4 = domain
    #
    my %f = (
        l    => [1],
        ln   => [ 1, 2 ],
        ld   => [ 4, 1 ],
        lp   => [ 3, 1 ],
        lnp  => [ 3, 1, 2 ],
        ldn  => [ 4, 1, 2 ],
        ldp  => [ 4, 3, 1 ],
        ldnp => [ 4, 3, 1, 2 ],
    );

    my %files;

    my $dir    = getcwd;
    my $re_dir = $dir;
    $re_dir =~ s/\./\\./g;

    finddepth(
        sub {
            my $filename        = $File::Find::name;
            my $stored_filename = $filename;
            if ( $md5 ) {
                eval {
                    require Digest::MD5;
                    Digest::MD5->import( 'md5_hex' );
                };
                die "This feature requires Digest::MD5" if $@;
                $stored_filename = md5_hex( $filename );
            }
            $filename =~ s/^$dir\///g;
            for ( @ignores ) {
                return if $filename =~ /$_/;
            }
            if ( @only ) {
                my $found = 0;
                for ( @only ) {
                    $found = 1 if $filename =~ /$_/;
                }
                return unless $found;
            }
            my @fileparts = split( '\.', $File::Find::name );
            my $ext = pop @fileparts;
            $files{$File::Find::name} = [ $ext, $filename, $stored_filename ] if grep { $ext eq $_ } keys %e;
        },
        $dir
    );

    my @found;
    for my $file ( sort keys %files ) {
        my ( $ext, $filename, $stored_filename ) = @{ $files{$file} };
        my $type = $e{$ext};
        print STDERR $type . " => " . $file . "\n";
        return if -l $file and not -e readlink( $file );
        my $parses = Locale::Simple::Scraper::Parser->new( type => $type )->from_file( $file );
        my @file_things = map {
            {
                %{ result_from_params( $_->{args}, $f{ $_->{func} } ) },
                  line => $_->{line},
                  file => $stored_filename,
                  type => $type,
            }
        } @{$parses};
        push @found, @file_things;
    }

    if ( $output eq 'po' ) {
        my %files;
        my %token;
        for ( @found ) {
            my $key .= defined $_->{domain}       ? '"' . $_->{domain} . '"'       : 'undef';
            $key    .= defined $_->{msgctxt}      ? '"' . $_->{msgctxt} . '"'      : 'undef';
            $key    .= defined $_->{msgid}        ? '"' . $_->{msgid} . '"'        : 'undef';
            $key    .= defined $_->{msgid_plural} ? '"' . $_->{msgid_plural} . '"' : 'undef';
            $token{$key} = $_ unless defined $token{$key};
            $files{$key} = [] unless defined $files{$key};
            push @{ $files{$key} }, $_->{file} . ':' . $_->{line};
        }
        for my $k ( sort { $a cmp $b } keys %files ) {
            print "\n";
            print "#: " . join( ' ', @{ $files{$k} } ) . "\n" if !$no_line_numbers;
            print "#, locale-simple-format";
            print " " . $token{$k}{domain} if defined $token{$k}{domain};
            print "\n";
            for ( qw( msgctxt msgid msgid_plural ) ) {
                print $_. ' "' . Locale::Simple::gettext_escape( $token{$k}{$_} ) . '"' . "\n"
                  if defined $token{$k}{$_};
            }
            my $plural_marker = $token{$k}{msgid_plural} ? "[0]" : "";
            print qq[msgstr$plural_marker ""\n];

        }
    }
    elsif ( $output eq 'perl' ) {
        print Dumper \@found;
    }
    elsif ( $output eq 'json' ) {
        eval {
            require JSON;
            JSON->import;
            print encode_json( \@found );
        } or do {
            die "You require the module JSON for this output";
        };
    }
    elsif ( $output eq 'yaml' ) {
        eval {
            require YAML;
            YAML->import;
            print Dump( \@found );
        } or do {
            die "You require the module YAML for this output";
        };
    }
}

sub parse_line {
    my ( $line, $type, $f, @results ) = @_;
    return if $line =~ /^\s*\#.*/;
    for ( keys %{$f} ) {
        my @args = @{ $f->{$_} };
        my $params = get_func_params( $_, $line );
        next if !$params;
        my $argc = scalar @args;
        my ( $remainder, @params ) = parse_params( $params, $type, $argc );
        if ( scalar @params == $argc ) {
            push @results, result_from_params( \@params, \@args ), parse_line( $remainder, $type, $f );
        }
    }
    return @results;
}

sub result_from_params {
    my ( $params, $args ) = @_;
    my %result;
    my $pos = 0;
    for ( @{$args} ) {
        $result{msgid}        = $params->[$pos] if $_ eq 1;
        $result{msgid_plural} = $params->[$pos] if $_ eq 2;
        $result{msgctxt}      = $params->[$pos] if $_ eq 3;
        $result{domain}       = $params->[$pos] if $_ eq 4;
        $pos++;
    }
    return \%result;
}

sub get_func_params {
    my ( $func, $line ) = @_;
    $line =~ /([^\w]|^)$func\((.*)/;
    return $2;
}

sub parse_params {
    my ( $params, $type, $argc ) = @_;
    my @chars = split( '', $params );
    my @args;
    my $arg         = "";
    my $q_state     = 0;    # 0 = code, 1 = qoute, 2 = double qoute
    my $comma_state = 1;
    while ( defined( my $c = shift @chars ) ) {
        next if $c =~ /\s/ and !$q_state;
        if ( $q_state ) {
            if ( $c eq '\\' ) {
                my $esc = shift @chars;
                if ( $esc eq "'" or $esc eq '"' or $esc eq '\\' ) {
                    $arg .= $esc;
                }
                else {
                    warn "Unknown escape char '" . $esc . "'";
                }
            }
            elsif ( ( $c eq "'" and $q_state == 1 ) or ( $c eq '"' and $q_state == 2 ) ) {
                $q_state     = 0;
                $comma_state = 0;
                push @args, $arg;
                $arg = "";
                last if scalar @args == $argc;
            }
            else {
                $arg .= $c;
            }
        }
        else {
            if ( $c eq "'" or $c eq '"' ) {
                die "quote found where comma expected: " . $params unless $comma_state;
                $q_state = $c eq "'" ? 1 : 2;
            }
            elsif ( $c eq ',' ) {
                die "comma found after comma in code: " . $params if $comma_state;
                $comma_state = 1;
            }
            elsif ( $type eq 'js' ) {
                last;
            }
            else {
                last;
            }
        }
    }
    return join( '', @chars ), @args;
}

1;

__END__

=pod

=head1 NAME

Locale::Simple::Scraper - scraper to find translation tokens in a directory

=head1 VERSION

version 0.018

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>, Torsten Raudssus <torsten@raudss.us>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Locale/Simple/Scraper/Parser.pm
use strict;
use warnings;

package Locale::Simple::Scraper::Parser;
BEGIN {
  $Locale::Simple::Scraper::Parser::AUTHORITY = 'cpan:GETTY';
}
$Locale::Simple::Scraper::Parser::VERSION = '0.018';
# ABSTRACT: parser to finds translation tokens in a code file

use base qw( Parser::MGC );

use Moo;
use Try::Tiny;
use curry;

has func_qr => ( is => 'ro', default => sub { qr/\bl(|n|p|np|d|dn|dnp)\b/ } );
has found   => ( is => 'ro', default => sub { [] } );
has type => ( is => 'ro', required => 1 );

with "Locale::Simple::Scraper::ParserShortcuts";

sub parse {
    my ( $self ) = @_;
    $self->sequence_of( $self->c_any_of( $self->curry::noise, $self->curry::call ) );
    return $self->found;
}

sub noise {
    my ( $self ) = @_;
    my $noise = $self->substring_before( $self->func_qr );
    $self->fail( "no noise found" ) if !length $noise;
    $self->debug( "discarded %d characters of noise", length $noise );
    return $noise;
}

sub call {
    my ( $self ) = @_;

    my $func = $self->expect( $self->func_qr );
    my $line = ( $self->where )[0];
    $self->debug( "found func $func at line %d", $line );

    try {
        my $arguments = $self->arguments( $func );
        push @{ $self->found }, { func => $func, args => $arguments, line => $line };
    }
    catch {
        die $_ if !eval { $_->isa( "Parser::MGC::Failure" ) };
        $self->warn_failure( $_ );
    };

    return;
}

sub arguments {
    my ( $self, $func ) = @_;

    my @arguments = ( $self->op( "(" ), $self->required_args( $func ), $self->extra_arguments, $self->op( ")" ) );
    $self->debug( "found %d arguments", scalar @arguments );

    return \@arguments;
}

sub op {
    my ( $self, $op ) = @_;
    return if $self->with_ws( maybe_expect => qr/\s*\Q$op\E/ );
    $self->fail( "Expected \"$op\"" );
}

sub extra_arguments {
    my ( $self ) = @_;
    return if !$self->maybe_expect( "," );

    my @types = ( $self->curry::call, $self->curry::dynamic_string, $self->curry::token_int, $self->curry::variable );
    my $extra_args = $self->list_of( ",", $self->c_any_of( @types ) );
    return @{$extra_args};
}

sub required_args {
    my ( $self, $func ) = @_;
    my %arg_lists = (
        l    => [qw( tr_token )],
        ln   => [qw( tr_token    comma  plural_token  comma  plural_count )],
        lp   => [qw( context_id  comma  tr_token )],
        lnp  => [qw( context_id  comma  tr_token      comma  plural_token  comma  plural_count )],
        ld   => [qw( domain_id   comma  tr_token )],
        ldn  => [qw( domain_id   comma  tr_token      comma  plural_token  comma  plural_count )],
        ldnp => [qw( domain_id   comma  context_id    comma  tr_token      comma  plural_token  comma  plural_count )],
    );
    return $self->collect_from( $arg_lists{$func} );
}

sub tr_token     { shift->named_token( "translation token" ) }
sub plural_token { shift->named_token( "plural translation token" ) }
sub plural_count { shift->named_token( "count of plural entity", "token_int" ) }
sub context_id   { shift->named_token( "context id" ) }
sub domain_id    { shift->named_token( "domain id" ) }
sub comma        { shift->op( "," ) }
sub variable     { shift->expect( qr/[\w\.]+/ ) }

sub constant_string {
    my ( $self, @components ) = @_;

    my $p = $self->{patterns};

    unshift @components,
      $self->curry::scope_of( q["], $self->c_with_ws( "double_quote_string_contents" ), q["] ),
      $self->curry::scope_of( q['], $self->c_with_ws( "single_quote_string_contents" ), q['] );

    my $string = $self->list_of( $self->concat_op, $self->c_any_of( @components ) );

    return join "", map { $_ ? $_ : "" } @{$string} if @{$string};

    $self->fail;
}

sub concat_op {
    my %ops = ( js => "+", pl => ".", tx => qr/(_|~)/, py => "+" );
    return $ops{ shift->type };
}

sub dynamic_string {
    my ( $self ) = @_;
    return $self->constant_string( $self->curry::call, $self->curry::variable );
}

sub double_quote_string_contents {
    my ( $self ) = @_;
    return $self->string_contents( $self->c_expect( qr/[^\\"]+/ ), $self->c_expect_escaped( q["] ) );
}

sub single_quote_string_contents {
    my ( $self ) = @_;
    return $self->string_contents(
        $self->c_expect( qr/[^\\']+/ ),
        $self->c_expect_escaped( q['] ),
        $self->c_expect_escaped( q[\\] ),
        $self->c_expect( qr/\\/ ),
    );
}

sub string_contents {
    my ( $self, @contents ) = @_;
    my $elements = $self->sequence_of( $self->c_any_of( @contents ) );
    return join "", @{$elements} if @{$elements};
    $self->fail( "no string contents found" );
}

1;

__END__

=pod

=head1 NAME

Locale::Simple::Scraper::Parser - parser to finds translation tokens in a code file

=head1 VERSION

version 0.018

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>, Torsten Raudssus <torsten@raudss.us>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Locale/Simple/Scraper/ParserShortcuts.pm
use strict;
use warnings;

package Locale::Simple::Scraper::ParserShortcuts;
BEGIN {
  $Locale::Simple::Scraper::ParserShortcuts::AUTHORITY = 'cpan:GETTY';
}
$Locale::Simple::Scraper::ParserShortcuts::VERSION = '0.018';
# ABSTRACT: LSS::Parser methods that are generic shortcuts to P::MGC functionality

use Moo::Role;

has debug_sub => (
    is      => 'ro',
    default => sub {
        sub { shift, warn "- " . sprintf shift() . "\n", @_ }
    }
);

sub debug { shift->debug_sub->( @_ ) }

sub collect_from {
    my ( $self, $methods ) = @_;
    return map { $self->$_ } @{$methods};
}

sub named_token {
    my ( $self, $name, $type ) = @_;
    $type ||= "constant_string";
    my $token = $self->maybe( sub { $self->$type } ) or $self->fail( "Expected $name" );
    return $token;
}

sub c_expect_escaped {
    my ( $self, $char ) = @_;
    return sub {
        $self->expect( qr/\\\Q$char\E/ );
        return $char;
    };
}

sub warn_failure {
    my ( $self, $f ) = @_;
    my ( $linenum, $col, $text ) = $self->where( $f->{pos} || $self->pos );
    my $indent = substr( $text, 0, $col );
    $_ =~ s/\t/    /g for $text, $indent;
    $indent =~ s/./-/g;     # blank out all the non-whitespace
    $text   =~ s/\%/%%/g;
    $self->debug( "$f->{message}:\n |$text\n |$indent^" );
    return;
}

sub c_any_of {
    my ( $self, @args ) = @_;
    return sub {
        $self->any_of( @args );
    };
}

sub c_expect {
    my ( $self, @args ) = @_;
    return sub {
        $self->expect( @args );
    };
}

sub c_with_ws {
    my ( $self, $code, @args ) = @_;
    return sub {
        local $self->{patterns}{ws} = qr//;
        return $self->$code( @args );
    };
}

sub with_ws {
    my ( $self, $code, @args ) = @_;
    local $self->{patterns}{ws} = qr//;
    return $self->$code( @args );
}

1;

__END__

=pod

=head1 NAME

Locale::Simple::Scraper::ParserShortcuts - LSS::Parser methods that are generic shortcuts to P::MGC functionality

=head1 VERSION

version 0.018

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>, Torsten Raudssus <torsten@raudss.us>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Functions for translate text based on gettext data, also in JavaScript",
  "AUTHOR" => "Torsten Raudssus <torsten\@raudss.us>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Locale-Simple",
  "EXE_FILES" => [
    "bin/po2json",
    "bin/locale_simple_scraper_old",
    "bin/locale_simple_scraper"
  ],
  "LICENSE" => "perl",
  "NAME" => "Locale::Simple",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "Data::Dumper" => 0,
    "Digest::MD5" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "File::Find" => 0,
    "Getopt::Long" => 0,
    "IO::All" => "0.41",
    "JSON" => "2.53",
    "Locale::PO" => "0.21",
    "Locale::TextDomain" => "1.23",
    "Locale::gettext_dumb" => 0,
    "Moo" => 0,
    "Moo::Role" => 0,
    "POSIX" => 0,
    "Parser::MGC" => 0,
    "Try::Tiny" => 0,
    "YAML" => 0,
    "base" => 0,
    "curry" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::InDistDir" => 0,
    "Test::More" => "0.98",
    "Test::Regression" => 0,
    "utf8" => 0
  },
  "VERSION" => "0.018",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => 0,
  "Cwd" => 0,
  "Data::Dumper" => 0,
  "Digest::MD5" => 0,
  "Exporter" => 0,
  "File::Basename" => 0,
  "File::Find" => 0,
  "File::Temp" => 0,
  "FindBin" => 0,
  "Getopt::Long" => 0,
  "IO::All" => "0.41",
  "JSON" => "2.53",
  "Locale::PO" => "0.21",
  "Locale::TextDomain" => "1.23",
  "Locale::gettext_dumb" => 0,
  "Moo" => 0,
  "Moo::Role" => 0,
  "POSIX" => 0,
  "Parser::MGC" => 0,
  "Test::InDistDir" => 0,
  "Test::More" => "0.98",
  "Test::Regression" => 0,
  "Try::Tiny" => 0,
  "YAML" => 0,
  "base" => 0,
  "curry" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


